### PR TITLE
Source Linnworks: refactor stream StockLocations

### PR DESCRIPTION
**Pull request by monai**
## What

This PR refactors private stream `Location`, which was embedded into stream `StockLocation`, to public stream `StockLocationDetails` utilizing `HttpSubStream`. From now on, `StockLocations` can be synced independently, while `StockLocationDetails` performs single sync of `StockLocations`, resulting in N+1 requests.

## Recommended reading order
1. `streams.py`
2. `source.py`

## 🚨 User Impact 🚨

🚨🚨 This PR changes the stream `StockLocations` schema.

#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated 
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

### CODE REVIEW

File: airbyte-integrations/connectors/source-linnworks/unit_tests/test_streams.py [hunk 1]
```diff
@@ -68,22 +68,22 @@ def test_backoff_time(patch_base_class, requests_mock, header_name, header_value
...
+)
+def test_stock_locations_details_request_params(mocker, stream_slice, expected):
+    source = StockLocationDetails()
 
-    assert list(records) == [
-        {"StockLocationId": 1, "location": {"FakeLocationFor": {"pkStockLocationId": 1}}},
-        {"StockLocationId": 2, "location": {"FakeLocationFor": {"pkStockLocationId": 2}}},
-    ]
+    if stream_slice:
+        records = source.request_params(None, stream_slice)
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-linnworks/unit_tests/test_streams.py:82]:
```suggestion
        params = source.request_params(None, stream_slice)
```

File: airbyte-integrations/connectors/source-linnworks/source_linnworks/streams.py [hunk 2]
```diff
@@ -70,50 +69,37 @@ def next_page_token(self, response: requests.Response) -> Optional[Mapping[str,
...
-
-        for record in records:
-            location = Location(authenticator=self.authenticator)
-            stock_location_records = location.read_records(
-                sync_mode, cursor_field, stream_slice, {"pkStockLocationId": record["StockLocationId"]}
-            )
-            record["location"] = next(stock_location_records)
-            yield record
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-linnworks/source_linnworks/streams.py:100]:
```suggestion
        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any], **kwargs
```
As `stream_slice` is required I don't think you should default it to `None`

File: airbyte-integrations/connectors/source-linnworks/source_linnworks/streams.py [hunk 3]
```diff
@@ -70,50 +69,37 @@ def next_page_token(self, response: requests.Response) -> Optional[Mapping[str,
...
-    # https://apps.linnworks.net/Api/Method/Inventory-GetStockLocations
-    # Response: List<StockLocation> https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Inventory-ClassBase-StockLocation
+class StockLocationDetails(HttpSubStream, StockLocations):
+    # https://apps.linnworks.net/Api/Method/Locations-GetLocation
+    # Response: StockLocation https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Locations-ClassBase-StockLocation
     # Allows 150 calls per minute
     primary_key = "StockLocationIntId"
 
+    def __init__(self, **kwargs):
+        super().__init__(StockLocations(**kwargs), **kwargs)
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-linnworks/source_linnworks/streams.py:92]: You could unit test the `__init__` to make sure future eventual changes to the parent declaration makes the test fail.

**Review by alafanechere — APPROVED** [review]
Thanks @monai for this cool refacto, I made minor suggestions + committed the docker image version bump. Let me know if you plan to make the suggested modification, I'll publish the docker image update from my side branch.

**Comment by monai** [comment]
> Thanks @monai for this cool refacto, I made minor suggestions + committed the docker image version bump. Let me know if you plan to make the suggested modification, I'll publish the docker image update from my side branch.

Thank you for review. I'm going to make changes on monday.

**Comment by alafanechere** [comment]
Thanks for the changes and this clean contrib, ready for merge! 🎉

**Pull request merged**